server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -821,57 +809,50 @@ public void testCounterOnPrimary() throws Exception {
             @Override
-            protected ReplicationOperation<Request, Request, TransportReplicationAction.PrimaryResult<Request, TestResponse>>
-            createReplicatedOperation(
-                    Request request,
-                    ActionListener<TransportReplicationAction.PrimaryResult<Request, TestResponse>> actionListener,
-                    TransportReplicationAction<Request, Request, TestResponse>.PrimaryShardReference primaryShardReference) {
+            protected void shardOperationOnPrimary(Request shardRequest, IndexShard primary,
+                                                   ActionListener<PrimaryResult<Request, TestResponse>> listener) {
                 assertIndexShardCounter(1);
-                if (throwExceptionOnCreation) {
-                    throw new ElasticsearchException("simulated exception, during createReplicatedOperation");
+                if (throwExceptionOnRun) {
+                    throw new ElasticsearchException("simulated exception, during shardOperationOnPrimary");
+                } else if (respondWithError) {
+                    listener.onFailure(new ElasticsearchException("simulated exception, as a response"));
+                } else {
+                    super.shardOperationOnPrimary(request, primary, listener);
                 }
-                return new NoopReplicationOperation(request, actionListener, primaryTerm) {
-                    @Override
-                    public void execute() throws Exception {
-                        assertIndexShardCounter(1);
-                        assertPhase(task, "primary");
-                        if (throwExceptionOnRun) {
-                            throw new ElasticsearchException("simulated exception, during performOnPrimary");
-                        } else if (respondWithError) {
-                            this.resultListener.onFailure(new ElasticsearchException("simulated exception, as a response"));
-                        } else {
-                            super.execute();
-                        }
-                    }
-                };
             }
-        }.run();
+        }.new AsyncPrimaryAction(primaryRequest, listener, task).run();
+
         assertIndexShardCounter(0);
         assertTrue(listener.isDone());
         assertPhase(task, "finished");
 
         try {
             listener.get();
         } catch (ExecutionException e) {
-            if (throwExceptionOnCreation || throwExceptionOnRun || respondWithError) {
+            if (throwExceptionOnRun || respondWithError) {
                 Throwable cause = e.getCause();
                 assertThat(cause, instanceOf(ElasticsearchException.class));
                 assertThat(cause.getMessage(), containsString("simulated"));
+                return;
             } else {
                 throw e;
             }
         }
+
+        if (throwExceptionOnRun || respondWithError) {
```

Comment:
nit: I think it is more logical to put this inside the try-catch (after listener.get()) and remove the return above.

Reply:
Sure, I pushed fb1f7eb9ec1b473bbaa319c11f42bc827ec27848.